src/balance.rs

```diff
@@ -26,41 +41,222 @@ impl<T: Config> From<Error<T>> for StatusCode {
 
 /// Balance adapter for providing basic access to balance cheques within the MoveVM.
 #[derive(Clone)]
-pub struct BalanceAdapter<T: Config> {
+pub struct BalanceAdapter<T: Config + SysConfig>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     _pd_config: PhantomData<T>,
+    /// Copy and virtual balances record of involved users.
+    balances: Rc<RefCell<HashMap<AccountIdOf<T>, BalanceOf<T>>>>,
+    /// Copy of initial state, without tracking it.
+    initial_state: Rc<RefCell<HashMap<AccountIdOf<T>, BalanceOf<T>>>>,
 }
 
-impl<T: Config> BalanceAdapter<T> {
+impl<T: Config + SysConfig> BalanceAdapter<T>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     /// Create a new [`BalanceAdapter`].
     pub fn new() -> Self {
         BalanceAdapter {
             _pd_config: PhantomData,
+            balances: Rc::new(RefCell::new(HashMap::new())),
+            initial_state: Rc::new(RefCell::new(HashMap::new())),
+        }
+    }
+
+    /// Inserts an initial balance state, e.g. the executor of a script.
+    pub fn insert_initial_balance(
+        &self,
+        account: &AccountIdOf<T>,
+        balance: &BalanceOf<T>,
+    ) -> DispatchResult {
+        if self.balances.borrow().contains_key(account)
+            || self.initial_state.borrow().contains_key(account)
+        {
+            return Err(DispatchError::Other("initial balance already inserted"));
+        }
```

Comment:
There is no need for this error - we should allow someone to call `write_cheque` (currently named `insert_initial_balance`) multiple times (although we probably won't need it, maybe in multisignature setup)

So, this scenario should be valid:
```rust
balance.write_cheque(account1, 45);
balance.write_cheque(account1, 55);

assert_eq!(balance.cheque_amount(account1), 100);
```

Reply:
Mhh, maybe we have to huddle about this :laughing: My intention of this method was to setup the initial available balance on the cheque, which shall present a limit of tokens to be used in the script execution. Why should it be updated multiple times?

Reply:
Done :+1: